crates/bevy_pbr/src/meshlet/asset.rs

```diff
@@ -68,84 +70,141 @@ pub struct MeshletBoundingSpheres {
-        }
-
-        let mut bytes = Vec::new();
-        reader.read_to_end(&mut bytes).await?;
-        let asset = bincode::deserialize_from(FrameDecoder::new(Cursor::new(bytes)))?;
+        writer: &'a mut Writer,
+        asset: SavedAsset<'a, MeshletMesh>,
+        _settings: &'a (),
+    ) -> Result<(), MeshletMeshSaveOrLoadError> {
+        // Compress asset data
+        let mut compressed_asset_data = Vec::new();
+        let mut compressor = FrameEncoder::new(&mut compressed_asset_data);
+        write_slice(&asset.vertex_data, &mut compressor)?;
+        write_slice(&asset.vertex_ids, &mut compressor)?;
+        write_slice(&asset.indices, &mut compressor)?;
+        write_slice(&asset.meshlets, &mut compressor)?;
+        write_slice(&asset.bounding_spheres, &mut compressor)?;
+        compressor.finish()?;
+
+        // Write asset metadata
+        writer
+            .write_all(&MESHLET_MESH_ASSET_VERSION.to_le_bytes())
+            .await?;
+        writer
+            .write_all(&asset.worst_case_meshlet_triangles.to_le_bytes())
+            .await?;
+        writer
+            .write_all(&(compressed_asset_data.len() as u64).to_le_bytes())
+            .await?;
 
-        Ok(asset)
-    }
+        // Write compressed asset data
+        writer.write_all(&compressed_asset_data).await?;
 
-    fn extensions(&self) -> &[&str] {
-        &["meshlet_mesh"]
+        Ok(())
     }
 }
 
-impl AssetSaver for MeshletMeshSaverLoad {
+impl AssetLoader for MeshletMeshSaverLoader {
     type Asset = MeshletMesh;
     type Settings = ();
-    type OutputLoader = Self;
     type Error = MeshletMeshSaveOrLoadError;
 
-    async fn save<'a>(
+    async fn load<'a>(
         &'a self,
-        writer: &'a mut Writer,
-        asset: SavedAsset<'a, Self::Asset>,
-        _settings: &'a Self::Settings,
-    ) -> Result<(), Self::Error> {
-        writer
-            .write_all(&MESHLET_MESH_ASSET_VERSION.to_le_bytes())
-            .await?;
+        reader: &'a mut dyn Reader,
+        _settings: &'a (),
+        _load_context: &'a mut LoadContext<'_>,
+    ) -> Result<MeshletMesh, MeshletMeshSaveOrLoadError> {
+        // Load and check asset version
+        let version = async_read_u64(reader).await?;
+        if version != MESHLET_MESH_ASSET_VERSION {
+            return Err(MeshletMeshSaveOrLoadError::WrongVersion { found: version });
+        }
 
-        let mut bytes = Vec::new();
-        let mut sync_writer = FrameEncoder::new(&mut bytes);
-        bincode::serialize_into(&mut sync_writer, asset.get())?;
-        sync_writer.finish()?;
-        writer.write_all(&bytes).await?;
+        // Load asset metadata
+        let worst_case_meshlet_triangles = async_read_u64(reader).await?;
+        let compressed_asset_data_len = async_read_u64(reader).await? as usize;
+
+        // Load compressed asset data
+        let mut compressed_asset_data = Vec::with_capacity(compressed_asset_data_len);
```

Comment:
I wish I could directly wrap Bevy's `&mut dyn Reader` in a FrameDecoder, rather than read to a Vec first, but it's not possible as Reader doesn't implement the typical std::io::Read sync trait - only Bevy's custom async traits :(

@cart any thoughts?

Reply:
We can make adapters like so:
```rust

struct AsyncWriteSyncAdapter<'a>(&'a mut Writer);

impl Write for AsyncWriteSyncAdapter<'_> {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        block_on(self.0.write(buf))
    }

    fn flush(&mut self) -> std::io::Result<()> {
        block_on(self.0.flush())
    }
}

struct AsyncReadSyncAdapter<'a>(&'a mut dyn Reader);

impl Read for AsyncReadSyncAdapter<'_> {
    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
        block_on(self.0.read(buf))
    }
}
```

Reply:
Not having to allocate an intermediate vec was a nice perf/memory win.